File: MainGame/FrontEndToolFS/HexPlane/HexValues.fs
namespace FrontEndToolFS.HexPlane

open System.IO

type HexValues =
    struct
        val values: int
        new(values: int) = { values = values }

        member this.Get mask shift =
            int (uint this.values >>> shift) &&& mask

        member this.With value mask shift =
            HexValues((this.values &&& ~~~(mask <<< shift)) ||| ((value &&& mask) <<< shift))

        member this.Elevation = (this.Get 31 0) - 15
        member this.WithElevation value = this.With (value + 15) 31 0
        member this.WaterLevel = this.Get 31 5
        member this.WithWaterLevel value = this.With value 31 5
        member this.UrbanLevel = this.Get 3 10
        member this.WithUrbanLevel value = this.With value 3 10
        member this.FarmLevel = this.Get 3 12
        member this.WithFarmLevel value = this.With value 3 12
        member this.PlantLevel = this.Get 3 14
        member this.WithPlantLevel value = this.With value 3 14
        member this.SpecialIndex = this.Get 255 16
        member this.WithSpecialIndex index = this.With index 255 16
        member this.TerrainTypeIndex = this.Get 255 24
        member this.WithTerrainTypeIndex index = this.With index 255 24

        member this.Save(writer: BinaryWriter) =
            writer.Write(byte this.TerrainTypeIndex)
            writer.Write(byte <| this.Elevation + 127)
            writer.Write(byte <| this.WaterLevel)
            writer.Write(byte <| this.UrbanLevel)
            writer.Write(byte <| this.FarmLevel)
            writer.Write(byte <| this.PlantLevel)
            writer.Write(byte <| this.SpecialIndex)

        static member Load (reader: BinaryReader) header =
            HexValues()
            |> (fun v -> v.WithTerrainTypeIndex << int <| reader.ReadByte())
            |> (fun v ->
                let mutable elevation = int <| reader.ReadByte()

                if header >= 4 then
                    elevation <- elevation - 127

                v.WithElevation elevation)
            |> (fun v -> v.WithWaterLevel << int <| reader.ReadByte())
            |> (fun v -> v.WithUrbanLevel << int <| reader.ReadByte())
            |> (fun v -> v.WithFarmLevel << int <| reader.ReadByte())
            |> (fun v -> v.WithPlantLevel << int <| reader.ReadByte())
            |> (fun v -> v.WithSpecialIndex << int <| reader.ReadByte())
    end

File: MainProject/Code.Domains/Models/ValueObjects/PlanetGenerates/HexValues.cs
using Godot;

namespace Domains.Models.ValueObjects.PlanetGenerates;

/// <summary>
/// Copyright 2022 Jasper Flick
/// 来源：Catlike Coding - Unity - Hex Map 教程
/// 源码地址：https://bitbucket.org/catlikecoding-projects/hex-map-project/src/2399393cdf64ad7d83eaff456f1207aa214356e2/Assets/Scripts/HexValues.cs?at=release%2F3.4.0
/// 由 ZeromaXHe 进行针对 Godot 球面六边形地图的改造
/// </summary>
public struct HexValues
{
	/// <summary>
	/// Seven values stored in 32 bits.
	/// TTTTTTTT SSSSSSSS PPFFUUWW WWWEEEEE.
	/// </summary>
	/// <remarks>Not readonly to support hot reloading in Unity.</remarks>
// #pragma warning disable IDE0044 // Add readonly modifier
    private int _values;
// #pragma warning restore IDE0044 // Add readonly modifier

    private readonly int Get(int mask, int shift) =>
		_values >>> shift & mask;

    private readonly HexValues With(int value, int mask, int shift) => new()
	{
		_values = (_values & ~(mask << shift)) | ((value & mask) << shift)
	};

	public readonly int Elevation => Get(31, 0);
	public readonly HexValues WithElevation(int value) => With(value, 31, 0);
	public readonly int WaterLevel => Get(31, 5);
	public readonly int ViewElevation => Mathf.Max(Elevation, WaterLevel);
	public readonly bool IsUnderwater => WaterLevel > Elevation;
	public readonly HexValues WithWaterLevel(int value) => With(value, 31, 5);
	public readonly int UrbanLevel => Get(3, 10);
	public readonly HexValues WithUrbanLevel(int value) => With(value, 3, 10);
	public readonly int FarmLevel => Get(3, 12);
	public readonly HexValues WithFarmLevel(int value) => With(value, 3, 12);
	public readonly int PlantLevel => Get(3, 14);
	public readonly HexValues WithPlantLevel(int value) => With(value, 3, 14);
	public readonly int SpecialIndex => Get(255, 16);
	public readonly HexValues WithSpecialIndex(int index) => With(index, 255, 16);
	public readonly int TerrainTypeIndex => Get(255, 24);
	public readonly HexValues WithTerrainTypeIndex(int index) => With(index, 255, 24);
}